tests/test_unittest.py

```diff
@@ -764,3 +776,43 @@ def test_penalty_time_constraint(value):
         np.array([]),
     )
     np.testing.assert_almost_equal(ocp.nlp[0].g_bounds[0], np.array([]))
+
+
+def custom(ocp, nlp, t, x, u, p, **extra_params):
+    states_idx = PenaltyFunctionAbstract._check_and_fill_index((), nlp.nx, "states_idx")
```

Comment:
If this is the reason PenaltyFunctionAbstract was added, this can be replaced by one of the following (Objective.Lagrange, Objective.Mayer, Constraint). If you want to choose the right one for the job (that is choosing Objective.Lagrange when dealing with Objective.Lagrange), you can pass `penalty_origin` as extra_params